File: programmers/stack_and_queue/Lv2_기능개발.py
import math
from collections import deque


def solution(progresses, speeds):
    answer = []
    days = deque()
    progresses = deque(progresses)
    speeds = deque(speeds)

    while speeds:
        s, p = speeds.popleft(), progresses.popleft()
        left = 100- p
        days.append(math.ceil(left /s))
    print(days)
    cnt = 1
    while len(days) > 1:
        d = days.popleft()
        while days and d >= days[0]:
            days.popleft()
            cnt += 1
        answer.append(cnt)
        cnt = 1
    if days:
        answer.append(1)

    return answer


print(solution(	[93, 30, 55], [1, 30, 5])) #[2,1]
print(solution([95, 90, 99, 99, 80, 99], [1, 1, 1, 1, 1, 1]))   # 	[1, 3, 2]
print(solution([96, 99, 98, 97], [1, 1, 1, 1] ))  #[4]

File: programmers/stack_and_queue/Lv2_기능개발.swift
//
//  Lv2_기능개발.swift
//  
//
//  Created by Johyeon Yoon on 2021/06/26.
//

import Foundation

func solution(_ progresses:[Int], _ speeds:[Int]) -> [Int] {
    var answer : [Int] = []
    
    let leftover = progresses.map {100 - $0}
    var leftDays : [Int] = []
    for i in 0..<speeds.count {
        leftDays.append(Int(ceil(Double(leftover[i])/Double(speeds[i]))))
    }
    
    var temp = leftDays[0]
    var cnt = 1
    for idx in 1..<leftDays.count {
        if leftDays[idx] > temp {  // 완성되는 다른 날에 처리가능함
            answer.append(cnt)
            cnt = 1
            temp = leftDays[idx]
        }
        else {
            cnt += 1
        }
        if idx == leftDays.count - 1 {
            answer.append(cnt)
        }
    }
    return answer
}

solution([95, 90, 99, 99, 80, 99], [1, 1, 1, 1, 1, 1])   //     [1, 3, 2]
